### PR TITLE
[Feature] FE interface for automatically creating partitions

### LINKED ISSUE #11444: Improve partition management
## Enhancement
Many databases support the function of importing and automatically creating partitions. We can support this function to complete the creation of partitions, and then use ttl to control the deletion of partitions. This ease of use will be better than the previous method.

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #11444

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [x] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java [hunk 1]
```diff
@@ -1436,6 +1471,210 @@ public TSetConfigResponse setConfig(TSetConfigRequest request) throws TException
...
+        String granularity = granularityLiteral.getStringValue();
+        for (String partitionValue : partitionValues) {
+            DateTimeFormatter beginDateTimeFormat;
+            LocalDateTime beginTime;
+            LocalDateTime endTime;
+
+            try {
+                beginDateTimeFormat = DateUtils.probeFormat(partitionValue);
+                beginTime = DateUtils.parseStringWithDefaultHSM(partitionValue, beginDateTimeFormat);
+                switch (granularity.toLowerCase()) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java:1563]: Should the data parsing code be extracting into a Util and test it well ?

File: fe/fe-core/src/main/java/com/starrocks/catalog/ExpressionRangePartitionInfo.java [hunk 2]
```diff
@@ -41,10 +41,9 @@ public ExpressionRangePartitionInfo(List<Expr> partitionExprs, List<Column> colu
         super(columns);
         Preconditions.checkState(partitionExprs != null);
         Preconditions.checkState(partitionExprs.size() > 0);
-        Preconditions.checkState(columns != null);
         Preconditions.checkState(partitionExprs.size() == columns.size());
         this.partitionExprs = partitionExprs;
-        this.isMultiColumnPartition = partitionExprs.size() > 0;
+        this.isMultiColumnPartition = true;
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/catalog/ExpressionRangePartitionInfo.java:46]: Why?

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java (hunk 1 above)
Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/service/FrontendServiceImpl.java]: You are right

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java [hunk 3]
```diff
@@ -541,4 +551,55 @@ public static String parseLiteralExprToDateString(LiteralExpr expr, int offset)
...
+                DistributionDesc distributionDesc = olapTable.getDefaultDistributionInfo().toDistributionDesc();
+
+                SingleRangePartitionDesc singleRangePartitionDesc =
+                        new SingleRangePartitionDesc(true, partitionName, partitionKeyDesc, partitionProperties);
+
+                AddPartitionClause addPartitionClause = new AddPartitionClause(singleRangePartitionDesc, distributionDesc,
+                        partitionProperties, false);
+                result.put(partitionName, addPartitionClause);
+            } catch (AnalysisException e) {
+                throw new AnalysisException(String.format("failed to analysis partition value:%s", partitionValue));
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/AnalyzerUtils.java:599]:
```suggestion
                throw new AnalysisException(String.format("failed to analyse partition value:%s", partitionValue));
```

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Comment by Astralidea** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 163 / 219 (74.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|85|127|66.93%|[1486, 1487, 1488, 1492, 1493, 1494, 1497, 1498, 1499, 1504, 1505, 1506, 1510, 1511, 1512, 1519, 1520, 1521, 1525, 1526, 1527, 1531, 1532, 1533, 1538, 1539, 1540, 1544, 1545, 1546, 1550, 1551, 1552, 1561, 1562, 1563, 1564, 1571, 1572, 1573, 1574, 1618]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|23|32|71.88%|[581, 582, 584, 585, 587, 588, 590, 609, 610]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|5|6|83.33%|[290]|
|:large_blue_circle:|com/starrocks/sql/ast/ExpressionPartitionDesc.java|31|35|88.57%|[120, 125, 126, 141]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RangePartitionDesc.java|1|1|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

**Comment by Astralidea** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 163 / 219 (74.43%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|85|127|66.93%|[1486, 1487, 1488, 1492, 1493, 1494, 1497, 1498, 1499, 1504, 1505, 1506, 1510, 1511, 1512, 1519, 1520, 1521, 1525, 1526, 1527, 1531, 1532, 1533, 1538, 1539, 1540, 1544, 1545, 1546, 1550, 1551, 1552, 1561, 1562, 1563, 1564, 1571, 1572, 1573, 1574, 1618]|
|:large_blue_circle:|com/starrocks/sql/analyzer/AnalyzerUtils.java|23|32|71.88%|[581, 582, 584, 585, 587, 588, 590, 609, 610]|
|:large_blue_circle:|com/starrocks/sql/analyzer/CreateTableAnalyzer.java|5|6|83.33%|[290]|
|:large_blue_circle:|com/starrocks/sql/ast/ExpressionPartitionDesc.java|31|35|88.57%|[120, 125, 126, 141]|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|10|10|100.00%|[]|
|:large_blue_circle:|com/starrocks/common/Config.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|7|7|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/ast/RangePartitionDesc.java|1|1|100.00%|[]|

**Comment by Astralidea** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java [hunk 4]
```diff
@@ -568,11 +568,17 @@ private PartitionDesc getPartitionDesc(StarRocksParser.PartitionDescContext cont
...
-            if (expr instanceof SlotRef) {
-                SlotRef slotRef = (SlotRef) expr;
-                columnList = ImmutableList.of(slotRef.getColumnName());
+            Expr expr = (Expr) visit(context.partitionExpression());
+            List<String> columnList = Lists.newArrayList();
+            if (expr instanceof FunctionCallExpr) {
+                FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
+                ArrayList<Expr> children = functionCallExpr.getChildren();
+                for (Expr child : children) {
+                    if (child instanceof SlotRef) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:577]: need to check child is not SlotRef? like year(a+1) ?

Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/AstBuilder.java:577]: The current version only supports simple expressions, and complex situations will be expanded later depending on the situation.

**Review by Youngwb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**